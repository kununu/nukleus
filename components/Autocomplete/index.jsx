import React, {PropTypes} from 'react';
import Autosuggest from 'react-autosuggest';
import Scroll from 'react-scroll';

import styles from './index.scss';

import getElementPositionY from '../../utils/elementPosition';
import isMobile from '../../utils/mobileDetection';
import {
  controlLabel,
  controlNote,
  errorStyles,
  formControl,
  hidden,
  formGroup,
  srOnly
} from '../index.scss';

export default class Autocomplete extends React.Component {
  static propTypes = {
    data: PropTypes.object,
    disabled: PropTypes.bool,
    error: PropTypes.string,
    id: PropTypes.string.isRequired,
    inputStyle: PropTypes.string,
    label: PropTypes.string.isRequired,
    labelHidden: PropTypes.bool,
    name: PropTypes.string.isRequired,
    noSuggestionText: PropTypes.string,
    onGetSuggestions: PropTypes.func,
    placeholder: PropTypes.string,
    query: PropTypes.object,
    required: PropTypes.bool,
    requiredLabel: PropTypes.string,
    scrollOffset: PropTypes.number,
    scrollTo: PropTypes.bool,
    submitOnEnter: PropTypes.bool,
    value: PropTypes.string
  };

  static defaultProps = {
    data: {},
    disabled: false,
    error: null,
    inputStyle: 'inline',
    labelHidden: false,
    noSuggestionText: 'No results found',
    onGetSuggestions: null,
    placeholder: '',
    query: {},
    required: false,
    requiredLabel: '',
    scrollOffset: 0,
    scrollTo: false,
    submitOnEnter: false,
    value: ''
  }

  state = {
    showError: false,
    showNoSuggestionsText: false,
    suggestions: this.props.data.items || [],
    value: ''
  };

  componentWillMount () {
    this.updateValue(this.props.query[this.props.name] || this.props.value || '');
  }

  componentWillReceiveProps (nextProps) {
    if (!this.needsUpdate(nextProps)) return;
    if (nextProps.error) this.showError();
    this.updateValue(this.props.query[this.props.name] || nextProps.value || '');
  }

  onChange = (event, {newValue}) => {
    this.setState({
      showNoSuggestionsText: true,
      value: newValue
    });
    this.hideError();
  }

  onFocus = () => {
    this.scrollToElement();
  }

  onBlur = () => {
    this.hideNoSuggestionsText();
  }

  onSuggestionsFetchRequested = ({value}) => {
    this.setState({
      suggestions: this.getSuggestions(value)
    });
  }

  onSuggestionsClearRequested = () => {
    this.setState({
      suggestions: []
    });
  };

  onSuggestionSelected = (e, {method}) => {
    this.setState({
      showNoSuggestionsText: false
    });

    if (method === 'enter' && !this.props.submitOnEnter) {
      e.preventDefault();
    }
  }

  getSuggestions = value => {
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;

    if (inputValue) {
      if (this.props.onGetSuggestions) {
        this.props.onGetSuggestions(inputValue);
        return this.props.data.items;
      }

      return this.props.data.items.filter(data => data.item.toLowerCase().slice(0, inputLength) === inputValue);
    }

    return [];
  }

  getSuggestionValue = suggestion => suggestion.item;

  hideNoSuggestionsText = () => {
    this.setState({
      showNoSuggestionsText: false
    });
  }

  showError () {
    this.setState({
      showError: true
    });
  }

  hideError () {
    this.setState({
      showError: false
    });
  }

  scrollToElement = () => {
    if (this.props.scrollTo && isMobile) {
      const elementPos = getElementPositionY(this.node, this.props.scrollOffset);
      const scroll = Scroll.animateScroll;
      scroll.scrollTo(elementPos);
    }
  }

  needsUpdate ({value, query}) {
    return (
      value !== this.props.value ||
      query !== this.props.query
    );
  }

  updateValue (value) {
    this.setState({value});
  }

  renderSuggestion = suggestion => <span>{suggestion.item}<span className={styles.suggestionInfo}>&nbsp;({suggestion.itemInfo})</span></span>;

  render () {
    const {
      data: {isFetching},
      disabled,
      error,
      label,
      labelHidden,
      id,
      name,
      noSuggestionText,
      placeholder,
      required,
      requiredLabel
    } = this.props;

    const {
      showError,
      showNoSuggestionsText,
      suggestions,
      value
    } = this.state;

    const inputProps = {
      className: formControl,
      disabled,
      id,
      name,
      onBlur: this.onBlur,
      onChange: this.onChange,
      onFocus: this.onFocus,
      placeholder,
      required,
      value
    };

    return (
      <div
        ref={node => this.node = node}
        className={`${formGroup} ${styles[this.props.inputStyle]}`}>

        {requiredLabel &&
          <span className={`${controlNote} ${styles.requiredLabel}`}>
            {requiredLabel}
          </span>
        }

        {labelHidden &&
          <span className={srOnly}>
            {label}
          </span>
        }

        <label
          className={`${controlLabel} ${labelHidden && hidden}`}
          htmlFor={id}>
          {label}
        </label>

        <div className={styles.autoCompleteContainer}>
          <Autosuggest
            focusFirstSuggestion
            focusInputOnSuggestionClick={!isMobile}
            getSuggestionValue={this.getSuggestionValue}
            inputProps={inputProps}
            onSuggestionSelected={this.onSuggestionSelected}
            onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
            onSuggestionsClearRequested={this.onSuggestionsClearRequested}
            renderSuggestion={this.renderSuggestion}
            suggestions={suggestions}
            theme={styles} />

          {isFetching &&
            <span className={styles.spinner}>
              <i
                className="fa fa-spinner fa-pulse fa-3x fa-fw"
                aria-hidden="true" />
            </span>
          }

          {!isFetching && !suggestions.length && value && showNoSuggestionsText ?
            <div className={styles.suggestionsContainer}>
              <ul>
                <li className={styles.suggestion}>
                  {noSuggestionText}
                </li>
              </ul>
            </div>
            : ''
          }

          {showError &&
            <span className={errorStyles}>
              {error}
            </span>
          }
        </div>
      </div>
    );
  }
}
