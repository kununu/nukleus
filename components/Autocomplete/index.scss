@import '~kununu-theme/scss/base/variables';

.requiredLabel {
  position: absolute;
  bottom: 45px;
  right: 2px;
}

.spinner {
  position: absolute;
  right: 13px;
  top: 10px;
  color: $gray-base-50;

  i {
    font-size: $font-size-lg;
  }
}

.suggestionInfo {
  color: $gray-base-40;
  font-size: $font-size-sm;
  display: inline-block;
}

.inline {
  position: relative;

  @media (min-width: $screen-xs-min) {
    display: flex;
    flex-flow: row;
    align-items: center;
  }

  label {
    margin-right: 10px;
    margin-bottom: 0;
    width: 140px;
  }
}

.suggestionsContainer {
  box-shadow: 0 2px 10px rgba($black, .2);
  border-radius: $border-radius-base;
  background-clip: padding-box;
  font-size: $font-size-base;
  width: 100%;
  margin-top: 1px;
  background-color: $white;
  max-height: 300px;
  overflow: auto;
  position: absolute;
  top: 100%;
  left: 0;
  right: 0;
  z-index: 2;

  @media (max-width: $screen-xs-max) {
    width: auto;
    margin-left: -30px;
    margin-right: -30px;
    border-radius: 0;
  }

  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }
}

.suggestion {
  box-sizing: border-box;
  text-align: left;
  padding: 8px 15px;
}

.suggestionFocused {
  background: $blue;
  color: $white;
  cursor: pointer;

  .suggestionInfo {
    color: $white;
  }
}

.suggestionsFooter {
  border-top: 1px solid $gray-base-03;
  padding: 8px 12px;
}

.autoCompleteContainer {
  position: relative;
  width: 100%;

  * {
    box-sizing: border-box;
  }
}

// Inline error definition
// Consider abstracting out to another component if we start using this a lot
.error {
  border-radius: 4px;
  font-size: 11px;
  padding: 5px 12px;
  text-transform: uppercase;
  font-weight: $font-weight-regular;
  background-color: $brand-danger;
  color: $white;
  position: absolute;
  z-index: 1;
  left: 0;
  top: 45px;

  &[href] {
    &:hover,
    &:focus {
      background-color: darken($brand-danger, 10%);
    }
  }

  &:after {
    content: '';
    position: absolute;
    display: block;
    top: -5px;
    right: 45%;
    bottom: auto;
    width: 0;
    height: 0;
    border-style: solid;
    border-width: 0 6px 6px;
    border-color: transparent transparent $brand-danger;
  }
}
