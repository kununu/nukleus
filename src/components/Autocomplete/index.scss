@import '~@kununu/kununu-theme-v2/scss/base/variables';

.spinner {
  position: absolute;
  right: 13px;
  top: 11px;

  svg {
    width: 18px;
    height: 18px;
    fill: $gray-base-50;
    vertical-align: middle;
    animation: spin 2s infinite linear;
  }
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(359deg);
  }
}

.suggestionInfo {
  color: $gray-base-40;
  font-size: $font-size-sm;
  display: inline-block;
}

.suggestionsContainer {
  box-shadow: 0 2px 10px rgba($black, .2);
  border-radius: $border-radius-base;
  background-clip: padding-box;
  font-size: $font-size-base;
  width: 100%;
  margin-top: 1px;
  background-color: $white;
  max-height: 300px;
  overflow: auto;
  position: absolute;
  top: 100%;
  left: 0;
  right: 0;
  z-index: 2;

  @media (max-width: $screen-xs-max) {
    width: auto;
    border-radius: 0;
  }

  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }
}

.suggestion {
  box-sizing: border-box;
  text-align: left;
  padding: 8px 15px;
}

.suggestionHighlighted {
  background: $blue;
  color: $white;
  cursor: pointer;

  .suggestionInfo {
    color: $white;
  }
}

.suggestionsFooter {
  border-top: 1px solid $gray-base-03;
  padding: 8px 12px;
}

.autocompleteContainer {
  position: relative;
  width: 100%;

  * {
    box-sizing: border-box;
  }
}