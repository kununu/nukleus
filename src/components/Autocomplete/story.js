import React from 'react';
import {storiesOf} from '@storybook/react';
import {defaultOptions} from 'defaultOptions'; // eslint-disable-line import/no-unresolved
import {withOptions} from '@storybook/addon-options';
import {action} from '@storybook/addon-actions';
import {withMarkdownNotes} from '@storybook/addon-notes';
import {withInfo} from '@storybook/addon-info';
import {withKnobs, text} from '@storybook/addon-knobs/react';

import ThemeProvider from '../ThemeProvider';

import customTheme from './customTheme.scss';

import Autocomplete from './index';


const staticData = {
  items: [
    {item: 'apple', itemInfo: 'US'},
    {item: 'alpha', itemInfo: 'Vienna'},
    {item: 'IBM', itemInfo: 'US'},
    {item: 'kununu', itemInfo: 'Vienna'},
    {item: 'kununu', itemInfo: 'US'},
  ],
};

storiesOf('Autocomplete', module)
  .addDecorator(withOptions({...defaultOptions}))
  .addDecorator(withKnobs)
  .addDecorator((story, context) => withInfo('The Autocomplete can be used with static or dynamic data and works especially well with state librarys such as redux. To easily customize your spinner you can add pass an element as a spinner prop! To fetch data dynamically you should use the onGetSuggestions param to update the data being passed to the data component. <br /> To enable auto scroll on focus you need to use scrollTo and set a scrollOffset')(story)(context))
  .add('With static data', withMarkdownNotes('The Autocomplete can be used with static or dynamic data and works especially well with state librarys such as redux. To easily customize your spinner you can add pass an element as a spinner prop! To fetch data dynamically you should use the onGetSuggestions param to update the data being passed to the data component. <br /> To enable auto scroll on focus you need to use scrollTo and set a scrollOffset')(() => (
    <div>
      <h1>
        Autocomplete
      </h1>
      <hr />
      <div className="container pull-left">
        <div className="row">

          <h3>
            Autocomplete input with inline label:
          </h3>

          <Autocomplete
            autofocus
            data={staticData}
            id="autocompletes"
            label={text('label', 'Autocomplete')}
            name="autocomplete"
            placeholder={text('placeholder', 'Type something...')}
            onSelectSuggestion={action('onSelectSuggestion')}
            noSuggestionText={text('noSuggestionText', 'No results found')}
            onFocus={action('onFocus')}
            onBlur={action('onBlur')}
            scrollOffset={70}
            scrollTo
          />

          <br />

          <h3>
          Required Autocomplete input with block label:
          </h3>

          <Autocomplete
            inputStyle="block"
            data={staticData}
            id="autocompletes"
            label={text('label', 'Autocomplete')}
            name="autocomplete"
            placeholder={text('placeholder', 'Type something...')}
            isRequired
            onSelectSuggestion={action('onSelectSuggestion')}
            noSuggestionText={text('noSuggestionText', 'No results found')}
            onFocus={action('onFocus')}
            onBlur={action('onBlur')}
            requiredLabel={text('requiredLabel', 'required')}
            scrollOffset={70}
            scrollTo
          />
        </div>
      </div>
    </div>
  )))
  .add('custom', () => (
    <ThemeProvider theme={customTheme}>
      <h3>Custom theme for buttons</h3>

      <p>
        Here is an example of an autocomplete with a custom theme.
      </p>

      <Autocomplete
        inputStyle="block"
        data={staticData}
        id="autocompletes"
        label={text('label', 'Autocomplete')}
        name="autocomplete"
        placeholder={text('placeholder', 'Type something...')}
        isRequired
        onSelectSuggestion={action('onSelectSuggestion')}
        noSuggestionText={text('noSuggestionText', 'No results found')}
        onFocus={action('onFocus')}
        onBlur={action('onBlur')}
        requiredLabel={text('requiredLabel', 'required')}
        scrollOffset={70}
        scrollTo
      />
    </ThemeProvider>
  ));
