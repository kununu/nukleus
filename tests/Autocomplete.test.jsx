import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import Autocomplete from 'Autocomplete'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies

const staticAutocomplete = (
  <Autocomplete
    data={{
      items: [
        {item: 'apple', itemInfo: 'US'},
        {item: 'alpha', itemInfo: 'Vienna'},
        {item: 'IBM', itemInfo: 'US'},
        {item: 'kununu', itemInfo: 'Vienna'},
        {item: 'kununu', itemInfo: 'US'}
      ]
    }}
    value="test"
    disabled
    isRequired
    requiredLabel="Required"
    id="autocompletes"
    label="Autocomplete"
    name="autocomplete"
    placeholder="Type something..."
    scrollOffset={70}
    scrollTo />
);

test('Renders Autocomplete without crashing', () => {
  const component = renderer.create(staticAutocomplete);
  expect(component.toJSON()).toMatchSnapshot();
});

test('Renders no suggestions container', () => {
  const component = mount(staticAutocomplete);
  component.find('input').simulate('change', {target: {value: 'z'}});
  expect(toJson(component)).toMatchSnapshot();
});

test('Renders suggestions container', () => {
  const component = mount(staticAutocomplete);
  component.find('input').simulate('change', {target: {value: 'a'}});
  component.find('input').simulate('focus');
  expect(toJson(component)).toMatchSnapshot();
});

test('Hides no suggestions on blur', () => {
  const component = mount(staticAutocomplete);
  component.find('input').simulate('change', {target: {value: 'z'}});
  component.find('input').simulate('focus');
  component.find('input').simulate('blur');
  expect(toJson(component)).toMatchSnapshot();
});

test('Updates value on selection', () => {
  const component = mount(staticAutocomplete);
  component.find('input').simulate('change', {target: {value: 'a'}});
  component.find('input').simulate('focus');
  component.find('#react-autowhatever-1--item-0').simulate('click');
  expect(toJson(component)).toMatchSnapshot();
});

test('Fetches suggestions on change', () => {
  const spyFunc = jest.fn();
  const component = mount(
    <Autocomplete
      data={{items: []}}
      onGetSuggestions={spyFunc}
      id="autocompletes"
      label="Autocomplete"
      name="autocomplete" />
  );

  component.find('input').simulate('change', {target: {value: 'a'}});
  expect(spyFunc).toHaveBeenCalled();
});
