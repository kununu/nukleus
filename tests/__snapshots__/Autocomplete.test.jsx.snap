exports[`test Hides no suggestions on blur 1`] = `
<Autocomplete
  data={
    Object {
      "items": Array [
        Object {
          "item": "apple",
          "itemInfo": "US",
        },
        Object {
          "item": "alpha",
          "itemInfo": "Vienna",
        },
        Object {
          "item": "IBM",
          "itemInfo": "US",
        },
        Object {
          "item": "kununu",
          "itemInfo": "Vienna",
        },
        Object {
          "item": "kununu",
          "itemInfo": "US",
        },
      ],
    }
  }
  disabled={true}
  error={null}
  id="autocompletes"
  inputStyle="inline"
  label="Autocomplete"
  name="autocomplete"
  noSuggestionText="No results found"
  placeholder="Type something..."
  query={Object {}}
  required={true}
  requiredLabel="Required"
  scrollOffset={70}
  scrollTo={true}
  value="test">
  <div
    className="formGroup inline">
    <span
      className="controlNote requiredLabel">
      Required
    </span>
    <label
      className="controlLabel undefined"
      htmlFor="autocompletes">
      Autocomplete
    </label>
    <div
      className="autoCompleteContainer">
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusFirstSuggestion={true}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        id="1"
        inputProps={
          Object {
            "className": "formControl",
            "disabled": true,
            "id": "autocompletes",
            "name": "autocomplete",
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
            "placeholder": "Type something...",
            "required": true,
            "value": "z",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={undefined {}}>
        <Autowhatever
          focusedItemIndex={null}
          focusedSectionIndex={null}
          getSectionItems={[Function]}
          id="1"
          inputProps={
            Object {
              "className": "formControl",
              "disabled": true,
              "id": "autocompletes",
              "name": "autocomplete",
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
              "onKeyDown": [Function],
              "placeholder": "Type something...",
              "required": true,
              "value": "z",
            }
          }
          itemProps={[Function]}
          items={Array []}
          multiSection={false}
          renderInputComponent={[Function]}
          renderItem={[Function]}
          renderItemData={
            Object {
              "query": "z",
            }
          }
          renderItemsContainer={[Function]}
          renderSectionTitle={[Function]}
          shouldRenderSection={[Function]}
          theme={Object {}}>
          <div
            style={Object {}}>
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={false}
              aria-haspopup={false}
              aria-owns="react-autowhatever-1"
              autoComplete="off"
              className="formControl"
              disabled={true}
              id="autocompletes"
              name="autocomplete"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Type something..."
              required={true}
              role="combobox"
              style={Object {}}
              type="text"
              value="z" />
            <div
              id="react-autowhatever-1"
              style={Object {}} />
          </div>
        </Autowhatever>
      </Autosuggest>
    </div>
  </div>
</Autocomplete>
`;

exports[`test Renders Autocomplete without crashing 1`] = `
<div
  className="formGroup inline">
  <span
    className="controlNote requiredLabel">
    Required
  </span>
  <label
    className="controlLabel undefined"
    htmlFor="autocompletes">
    Autocomplete
  </label>
  <div
    className="autoCompleteContainer">
    <div
      style={Object {}}>
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-expanded={false}
        aria-haspopup={false}
        aria-owns="react-autowhatever-1"
        autoComplete="off"
        className="formControl"
        disabled={true}
        id="autocompletes"
        name="autocomplete"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Type something..."
        required={true}
        role="combobox"
        style={Object {}}
        type="text"
        value="test" />
      <div
        id="react-autowhatever-1"
        style={Object {}} />
    </div>
    
  </div>
</div>
`;

exports[`test Renders no suggestions container 1`] = `
<Autocomplete
  data={
    Object {
      "items": Array [
        Object {
          "item": "apple",
          "itemInfo": "US",
        },
        Object {
          "item": "alpha",
          "itemInfo": "Vienna",
        },
        Object {
          "item": "IBM",
          "itemInfo": "US",
        },
        Object {
          "item": "kununu",
          "itemInfo": "Vienna",
        },
        Object {
          "item": "kununu",
          "itemInfo": "US",
        },
      ],
    }
  }
  disabled={true}
  error={null}
  id="autocompletes"
  inputStyle="inline"
  label="Autocomplete"
  name="autocomplete"
  noSuggestionText="No results found"
  placeholder="Type something..."
  query={Object {}}
  required={true}
  requiredLabel="Required"
  scrollOffset={70}
  scrollTo={true}
  value="test">
  <div
    className="formGroup inline">
    <span
      className="controlNote requiredLabel">
      Required
    </span>
    <label
      className="controlLabel undefined"
      htmlFor="autocompletes">
      Autocomplete
    </label>
    <div
      className="autoCompleteContainer">
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusFirstSuggestion={true}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        id="1"
        inputProps={
          Object {
            "className": "formControl",
            "disabled": true,
            "id": "autocompletes",
            "name": "autocomplete",
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
            "placeholder": "Type something...",
            "required": true,
            "value": "z",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={undefined {}}>
        <Autowhatever
          focusedItemIndex={null}
          focusedSectionIndex={null}
          getSectionItems={[Function]}
          id="1"
          inputProps={
            Object {
              "className": "formControl",
              "disabled": true,
              "id": "autocompletes",
              "name": "autocomplete",
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
              "onKeyDown": [Function],
              "placeholder": "Type something...",
              "required": true,
              "value": "z",
            }
          }
          itemProps={[Function]}
          items={Array []}
          multiSection={false}
          renderInputComponent={[Function]}
          renderItem={[Function]}
          renderItemData={
            Object {
              "query": "z",
            }
          }
          renderItemsContainer={[Function]}
          renderSectionTitle={[Function]}
          shouldRenderSection={[Function]}
          theme={Object {}}>
          <div
            style={Object {}}>
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={false}
              aria-haspopup={false}
              aria-owns="react-autowhatever-1"
              autoComplete="off"
              className="formControl"
              disabled={true}
              id="autocompletes"
              name="autocomplete"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Type something..."
              required={true}
              role="combobox"
              style={Object {}}
              type="text"
              value="z" />
            <div
              id="react-autowhatever-1"
              style={Object {}} />
          </div>
        </Autowhatever>
      </Autosuggest>
      <div
        className="suggestionsContainer">
        <ul>
          <li
            className="suggestion">
            No results found
          </li>
        </ul>
      </div>
    </div>
  </div>
</Autocomplete>
`;

exports[`test Renders suggestions container 1`] = `
<Autocomplete
  data={
    Object {
      "items": Array [
        Object {
          "item": "apple",
          "itemInfo": "US",
        },
        Object {
          "item": "alpha",
          "itemInfo": "Vienna",
        },
        Object {
          "item": "IBM",
          "itemInfo": "US",
        },
        Object {
          "item": "kununu",
          "itemInfo": "Vienna",
        },
        Object {
          "item": "kununu",
          "itemInfo": "US",
        },
      ],
    }
  }
  disabled={true}
  error={null}
  id="autocompletes"
  inputStyle="inline"
  label="Autocomplete"
  name="autocomplete"
  noSuggestionText="No results found"
  placeholder="Type something..."
  query={Object {}}
  required={true}
  requiredLabel="Required"
  scrollOffset={70}
  scrollTo={true}
  value="test">
  <div
    className="formGroup inline">
    <span
      className="controlNote requiredLabel">
      Required
    </span>
    <label
      className="controlLabel undefined"
      htmlFor="autocompletes">
      Autocomplete
    </label>
    <div
      className="autoCompleteContainer">
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusFirstSuggestion={true}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        id="1"
        inputProps={
          Object {
            "className": "formControl",
            "disabled": true,
            "id": "autocompletes",
            "name": "autocomplete",
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
            "placeholder": "Type something...",
            "required": true,
            "value": "a",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={
          Array [
            Object {
              "item": "apple",
              "itemInfo": "US",
            },
            Object {
              "item": "alpha",
              "itemInfo": "Vienna",
            },
          ]
        }
        theme={undefined {}}>
        <Autowhatever
          focusedItemIndex={0}
          focusedSectionIndex={null}
          getSectionItems={[Function]}
          id="1"
          inputProps={
            Object {
              "className": "formControl",
              "disabled": true,
              "id": "autocompletes",
              "name": "autocomplete",
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
              "onKeyDown": [Function],
              "placeholder": "Type something...",
              "required": true,
              "value": "a",
            }
          }
          itemProps={[Function]}
          items={
            Array [
              Object {
                "item": "apple",
                "itemInfo": "US",
              },
              Object {
                "item": "alpha",
                "itemInfo": "Vienna",
              },
            ]
          }
          multiSection={false}
          renderInputComponent={[Function]}
          renderItem={[Function]}
          renderItemData={
            Object {
              "query": "a",
            }
          }
          renderItemsContainer={[Function]}
          renderSectionTitle={[Function]}
          shouldRenderSection={[Function]}
          theme={Object {}}>
          <div
            style={Object {}}>
            <input
              aria-activedescendant="react-autowhatever-1--item-0"
              aria-autocomplete="list"
              aria-expanded={true}
              aria-haspopup={true}
              aria-owns="react-autowhatever-1"
              autoComplete="off"
              className="formControl"
              disabled={true}
              id="autocompletes"
              name="autocomplete"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Type something..."
              required={true}
              role="combobox"
              style={Object {}}
              type="text"
              value="a" />
            <div
              id="react-autowhatever-1"
              style={Object {}}>
              <ItemsList
                focusedItemIndex={0}
                getItemId={[Function]}
                itemProps={[Function]}
                items={
                  Array [
                    Object {
                      "item": "apple",
                      "itemInfo": "US",
                    },
                    Object {
                      "item": "alpha",
                      "itemInfo": "Vienna",
                    },
                  ]
                }
                keyPrefix="react-autowhatever-1-"
                onFocusedItemChange={[Function]}
                renderItem={[Function]}
                renderItemData={
                  Object {
                    "query": "a",
                  }
                }
                sectionIndex={null}
                theme={[Function]}>
                <ul
                  role="listbox"
                  style={Object {}}>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={0}
                    id="react-autowhatever-1--item-0"
                    item={
                      Object {
                        "item": "apple",
                        "itemInfo": "US",
                      }
                    }
                    itemIndex={0}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "a",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={0}
                      id="react-autowhatever-1--item-0"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        apple
                        <span
                          className="suggestionInfo">
                           (
                          US
                          )
                        </span>
                      </span>
                    </li>
                  </Item>
                  <Item
                    data-section-index={null}
                    data-suggestion-index={1}
                    id="react-autowhatever-1--item-1"
                    item={
                      Object {
                        "item": "alpha",
                        "itemInfo": "Vienna",
                      }
                    }
                    itemIndex={1}
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onTouchStart={[Function]}
                    renderItem={[Function]}
                    renderItemData={
                      Object {
                        "query": "a",
                      }
                    }
                    sectionIndex={null}
                    style={Object {}}>
                    <li
                      data-section-index={null}
                      data-suggestion-index={1}
                      id="react-autowhatever-1--item-1"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                      role="option"
                      style={Object {}}>
                      <span>
                        alpha
                        <span
                          className="suggestionInfo">
                           (
                          Vienna
                          )
                        </span>
                      </span>
                    </li>
                  </Item>
                </ul>
              </ItemsList>
            </div>
          </div>
        </Autowhatever>
      </Autosuggest>
    </div>
  </div>
</Autocomplete>
`;

exports[`test Updates value on selection 1`] = `
<Autocomplete
  data={
    Object {
      "items": Array [
        Object {
          "item": "apple",
          "itemInfo": "US",
        },
        Object {
          "item": "alpha",
          "itemInfo": "Vienna",
        },
        Object {
          "item": "IBM",
          "itemInfo": "US",
        },
        Object {
          "item": "kununu",
          "itemInfo": "Vienna",
        },
        Object {
          "item": "kununu",
          "itemInfo": "US",
        },
      ],
    }
  }
  disabled={true}
  error={null}
  id="autocompletes"
  inputStyle="inline"
  label="Autocomplete"
  name="autocomplete"
  noSuggestionText="No results found"
  placeholder="Type something..."
  query={Object {}}
  required={true}
  requiredLabel="Required"
  scrollOffset={70}
  scrollTo={true}
  value="test">
  <div
    className="formGroup inline">
    <span
      className="controlNote requiredLabel">
      Required
    </span>
    <label
      className="controlLabel undefined"
      htmlFor="autocompletes">
      Autocomplete
    </label>
    <div
      className="autoCompleteContainer">
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusFirstSuggestion={true}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        id="1"
        inputProps={
          Object {
            "className": "formControl",
            "disabled": true,
            "id": "autocompletes",
            "name": "autocomplete",
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
            "placeholder": "Type something...",
            "required": true,
            "value": "apple",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={undefined {}}>
        <Autowhatever
          focusedItemIndex={null}
          focusedSectionIndex={null}
          getSectionItems={[Function]}
          id="1"
          inputProps={
            Object {
              "className": "formControl",
              "disabled": true,
              "id": "autocompletes",
              "name": "autocomplete",
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
              "onKeyDown": [Function],
              "placeholder": "Type something...",
              "required": true,
              "value": "apple",
            }
          }
          itemProps={[Function]}
          items={Array []}
          multiSection={false}
          renderInputComponent={[Function]}
          renderItem={[Function]}
          renderItemData={
            Object {
              "query": "apple",
            }
          }
          renderItemsContainer={[Function]}
          renderSectionTitle={[Function]}
          shouldRenderSection={[Function]}
          theme={Object {}}>
          <div
            style={Object {}}>
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={false}
              aria-haspopup={false}
              aria-owns="react-autowhatever-1"
              autoComplete="off"
              className="formControl"
              disabled={true}
              id="autocompletes"
              name="autocomplete"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Type something..."
              required={true}
              role="combobox"
              style={Object {}}
              type="text"
              value="apple" />
            <div
              id="react-autowhatever-1"
              style={Object {}} />
          </div>
        </Autowhatever>
      </Autosuggest>
    </div>
  </div>
</Autocomplete>
`;
